### PR TITLE
Total and current batch size

### PR DESCRIPTION
**Pull request by ouchadam**
Adds the total and current size of the batch to the `DownloadBatch` 

- removes the `CollatedDownloadInfo` in favour of reusing the `DownloadBatch`

### CODE REVIEW

File: library/src/main/java/com/novoda/downloadmanager/lib/DownloadNotifier.java [hunk 1]
```diff
@@ -167,27 +168,27 @@ private boolean shouldShowCompletedItem(int visibility) {
...
     }
 
-    private void addBatchToCluster(String tag, Map<String, List<DownloadBatch>> cluster, DownloadBatch batch) {
+    private void addBatchToCluster(String tag, Map<String, Collection<DownloadBatch>> cluster, DownloadBatch batch) {
         if (tag == null) {
             return;
         }
 
-        List<DownloadBatch> batches;
+        Collection<DownloadBatch> batches;
```
Comment (ronocod) [inline library/src/main/java/com/novoda/downloadmanager/lib/DownloadNotifier.java]: :+1:

**Comment by ronocod** [comment]
LGTM :+1:

File: library/src/main/java/com/novoda/downloadmanager/lib/DownloadNotifier.java [hunk 2]
```diff
@@ -167,27 +168,27 @@ private boolean shouldShowCompletedItem(int visibility) {
...
 
         if (cluster.containsKey(tag)) {
             batches = cluster.get(tag);
         } else {
-            batches = new ArrayList<>();
+            batches = new HashSet<>();
```
Comment (zegnus) [inline library/src/main/java/com/novoda/downloadmanager/lib/DownloadNotifier.java]: Why of this change? Do you need random access or guarantee free of duplication? HashSet will have a slower insertion than an ArrayList

File: library/src/main/java/com/novoda/downloadmanager/lib/DownloadNotifier.java [hunk 3]
```diff
@@ -167,27 +168,27 @@ private boolean shouldShowCompletedItem(int visibility) {
...
             cluster.put(tag, batches); // TODO not sure if this is right compared to ArrayListMultiMap
```
Comment (zegnus) [inline library/src/main/java/com/novoda/downloadmanager/lib/DownloadNotifier.java]: If you really want to nail it, remove the comment and replace the type of your cluster to `SimpleArrayMap`

Reference here https://medium.com/google-developers/developing-for-android-ii-bb9a51f8c8b9

Comment (ronocod) [inline library/src/main/java/com/novoda/downloadmanager/lib/DownloadNotifier.java]: Might need `ArrayMap` instead so that you can use the Iterator API: https://developer.android.com/reference/android/support/v4/util/ArrayMap.html

Comment (zegnus) [inline library/src/main/java/com/novoda/downloadmanager/lib/DownloadNotifier.java]: There is no need to use the iterator, you can always get the size and iterate manually

Comment (zegnus) [inline library/src/main/java/com/novoda/downloadmanager/lib/DownloadNotifier.java]: For example here

From
```java
for (String notificationId : clusters.keySet()) {
  int type = getNotificationTagType(notificationId);
```

to
```java
for (int i = 0, size = clusters.size(); i < size; i++) {
  String notificationId = clusters.keyAt(i);
  ...
}
```

Comment (ronocod) [inline library/src/main/java/com/novoda/downloadmanager/lib/DownloadNotifier.java]: We could do that yeah, it'd just make this a bit messier: https://github.com/novoda/download-manager/blob/total-and-current-batch-size/library/src/main/java/com/novoda/downloadmanager/lib/DownloadNotifier.java#L189

Comment (ronocod) [inline library/src/main/java/com/novoda/downloadmanager/lib/DownloadNotifier.java]: Commented at the same time :smile:

Comment (zegnus) [inline library/src/main/java/com/novoda/downloadmanager/lib/DownloadNotifier.java]: If we are going to do this for performance, I would go for the SimpleArrayMap and the manual iteration :), is not daaat messy :P

Comment (ronocod) [inline library/src/main/java/com/novoda/downloadmanager/lib/DownloadNotifier.java]: Yeah it doesn't look as bad as I thought actually

File: library/src/main/java/com/novoda/downloadmanager/lib/DownloadNotifier.java (hunk 2 above)
Comment (ouchadam) [inline library/src/main/java/com/novoda/downloadmanager/lib/DownloadNotifier.java]: doesn't need to be any more, it's been reverted

File: library/src/main/java/com/novoda/downloadmanager/lib/DownloadNotifier.java (hunk 3 above)
Comment (ouchadam) [inline library/src/main/java/com/novoda/downloadmanager/lib/DownloadNotifier.java]: it's been removed, sorry guys!

Comment (zegnus) [inline library/src/main/java/com/novoda/downloadmanager/lib/DownloadNotifier.java]: @ouchadam uh? This comments are about this line `cluster.put(tag, batches); // TODO not sure if this is right compared to ArrayListMultiMap` :P

Comment (ouchadam) [inline library/src/main/java/com/novoda/downloadmanager/lib/DownloadNotifier.java]: ahhhh I thought you were talking about the array list/hashset

Comment (ouchadam) [inline library/src/main/java/com/novoda/downloadmanager/lib/DownloadNotifier.java]: would prefer to leave the TODO for another pretty pleaseeeee

Comment (zegnus) [inline library/src/main/java/com/novoda/downloadmanager/lib/DownloadNotifier.java]: I'm OK moving this change into another PR so we can release another DM version right now

Comment (ronocod) [inline library/src/main/java/com/novoda/downloadmanager/lib/DownloadNotifier.java]: :+1: agreed

**Comment by zegnus** [comment]
:+1: LGTM with https://github.com/novoda/download-manager/pull/54#discussion_r33858809 on the queue

**Comment by ronocod** [comment]
I hear the release train approaching!

File: library/src/main/java/com/novoda/downloadmanager/lib/DownloadBatch.java [hunk 4]
```diff
@@ -2,18 +2,24 @@
 
 import java.util.List;
 
-class DownloadBatch {
+public class DownloadBatch {
```
Comment (blundell) [inline library/src/main/java/com/novoda/downloadmanager/lib/DownloadBatch.java]: in terms of library writing, we have now made this class available to clients, this means it's implementation is tied in to the version of the library.

I would have done a conversion from this internal represenation to something else `Collated..` and then THAT class can be responsible for keeping public API compatibility and this class can carry on being ruthlessly refactored.

I'd X this if it hadn't already been merged

File: library/src/main/java/com/novoda/downloadmanager/lib/DownloadClientReadyChecker.java [hunk 5]
```diff
@@ -5,9 +5,10 @@
 
     /**
      * This method is executed on a background thread
-     * @param collatedDownloadInfo
+     *
+     * @param downloadBatch
```
Comment (blundell) [inline library/src/main/java/com/novoda/downloadmanager/lib/DownloadClientReadyChecker.java]: don't add jdoc params unless you are going to fill them in with useful information

File: library/src/main/java/com/novoda/downloadmanager/lib/DownloadDeleter.java [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+    public DownloadDeleter(ContentResolver resolver) {
+        this.resolver = resolver;
+    }
+
+    public void deleteFileAndDatabaseRow(DownloadInfo info) {
+        deleteFileAndMediaReference(info);
+        resolver.delete(info.getAllDownloadsUri(), null, null);
+    }
+
+    public void deleteFileAndMediaReference(DownloadInfo info) {
```
Comment (blundell) [inline library/src/main/java/com/novoda/downloadmanager/lib/DownloadDeleter.java]: see the and in this method. That means internally it can look like this
```java
public void deleteFileAndMediaReference(DownloadInfo info) {
          deleteFile(info);
          deleteMediaReference(info);
}
```

FYI your method name should actually be `deleteMediaReferenceAndFile()`

If you had those two smaller methods, they could be easier to read with a *guard clause* & *prefer the positive*:

```java
public void deleteFileAndMediaReference(DownloadInfo info) {
     String mediaUri = info.mMediaProviderUri;
     if (TextUtils.isEmpty(mediaUri)) {
          return;
     }
     resolver.delete(Uri.parse(mediaUri), null, null);
}
```

File: library/src/main/java/com/novoda/downloadmanager/lib/DownloadDeleter.java [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+        }
+        if (info.mDestination != Downloads.Impl.DESTINATION_EXTERNAL && info.mFileName != null) {
+            Log.d("deleteDownloadLocked() deleting " + info.mFileName);
+            deleteFileIfExists(info.mFileName);
+        }
+        downloads.remove(info.mId);
+    }
+
+    private void deleteFileIfExists(String path) {
+        if (!TextUtils.isEmpty(path)) {
```
Comment (blundell) [inline library/src/main/java/com/novoda/downloadmanager/lib/DownloadDeleter.java]: if isEmpty return

File: library/src/main/java/com/novoda/downloadmanager/lib/DownloadClientReadyChecker.java (hunk 5 above)
Comment (ouchadam) [inline library/src/main/java/com/novoda/downloadmanager/lib/DownloadClientReadyChecker.java]: automerged

File: library/src/main/java/com/novoda/downloadmanager/lib/DownloadDeleter.java (hunk 6 above)
Comment (ouchadam) [inline library/src/main/java/com/novoda/downloadmanager/lib/DownloadDeleter.java]: this code was all moved but I see your point

Comment (blundell) [inline library/src/main/java/com/novoda/downloadmanager/lib/DownloadDeleter.java]: ah I see :+1:
